src/Client/HedgedConnections.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+#if defined(OS_LINUX)
+
+#include <functional>
+#include <queue>
+#include <Client/HedgedConnectionsFactory.h>
+#include <Client/IConnections.h>
+
+namespace DB
+{
+
+/** To receive data from multiple replicas (connections) from one shard asynchronously,
+  * The principe of Hedged Connections is used to reduce tail latency:
+  * (if we don't receive data from replica for a long time, we try to get new replica
+  * and send query to it, without cancelling working with previous replica). This class
+  * supports all functionality that MultipleConnections has.
+  */
+class HedgedConnections : public IConnections
+{
+public:
+    struct ReplicaState
+    {
+        Connection * connection = nullptr;
+        std::unordered_map<int, ConnectionTimeoutDescriptorPtr> active_timeouts;
+    };
+
+    struct ReplicaLocation
+    {
+        size_t offset;
+        size_t index;
+    };
+
+    struct OffsetState
+    {
+        std::vector<ReplicaState> replicas;
+        size_t active_connection_count;
+        bool first_packet_of_data_received;
```

Comment:
Looks like when `first_packet_of_data_received`, `active_connection_count` should always be one. This is an implicit invariant. It is good to add a comment about it. Also, if we have the invariant, we should consider if need `OffsetState` to be a class with private fields and some methods (I am not sure if it is reasonable for now).

Also, replicas are indexed by int, so `replicas.size()` is not always 1 when `first_packet_of_data_received`. However, we can probably replace `std::vector` to `std::list` and use `std::list::iterator` instead. (I also don't think it is a good idea so far)

Reply:
I thought about it, but I consider that making class with methods and using std::list will make this code more difficult